articles/working-with-apis-in-typescript/index.md

```diff
@@ -0,0 +1,296 @@
+
+```bash
+# Initializes a tsconfig file
+npx tsc --init
+```
+
+![TypeScript Configuration](/engineering-education/working-with-apis-in-typescript/tsconfig.png)
+
+Image: **TypeScript Configuration File**
+
+### Let's Code!
+
+#### Creating a configuration to store routes
+
+TypeScript is an Object-Oriented Programming (OOP) Language. An OOP language uses objects and classes to represent entities. You can read more about OOP [here](https://en.wikipedia.org/wiki/Object-oriented_programming).
+
+Let us write an abstract class that defines the skeleton for our route configuration. Create a new folder called `src`, and a file `routes.config.ts` within `src`:
+
+```bash
+mkdir src
+cd src
+
+# Create the routes config file
+touch routes.config.ts
+```
+
+**`routes.config.ts`**
+
+```ts
+// The express package is imported
+import express from "express";
+
+// We define this as an abstract class
+// This contains the abstract method: config()
+export abstract class RoutesConfig {
+  // app is of the TypeScript type: Application, which is defined in @types/express
+  app: express.Application;
+
+  // name is the name of the route
+  name: string;
+
+  constructor(app: express.Application, name: string) {
+    // Initializing the member variables
+    this.app = app;
+    this.name = name;
+    this.config();
+  }
+
+  // Name Getter Method
+  getName = () => {
+    return this.name;
+  };
+
+  // config() is defined as an abstract function
+  abstract config(): express.Application;
+}
+```
+
+You might wonder why the above class is defined as abstract. This is because abstract classes cannot contain the implementation of their methods. This routes configuration defines a basic skeleton and can be extended to create different routes.
+
+#### Defining the routes
+
+Now, let us define the various routes. Create a new file called `apiRoutes.ts`.
+
+```bash
+touch apiRoutes.ts
+```
+
+**`apiRoutes.ts`**
+
+```ts
+// The configuration file that we wrote is imported
+import { RoutesConfig } from "./routes.config";
+import express from "express";
+
+// The ApiRoutes class inherits the RoutesConfig class
+export class ApiRoutes extends RoutesConfig {
+  constructor(app: express.Application) {
+    super(app, "ApiRoutes");
+  }
+
+  // The config function contains the routes
+  config() {
+    // The express.Request and express.Response types are defined in @types/express
+    this.app
+      .route("/api")
+      .get((req: express.Request, res: express.Response) => {
+        // The response is sent back
+        res.send(`GET Request successful!`);
+      });
+    // :id defines a query param
+    this.app
+      .route("/api/:id")
+      .get((req: express.Request, res: express.Response) => {
+        // The parameter is sent back as a response
+        res.send(
+          `GET Request successful. Parameter passed is: ${req.params.id}`
+        );
+      });
+    return this.app;
```

Comment:
Apart from `GET` request, you could create 3 other APIs for `POST`, `PUT` and `DELETE` request, and explain about them. It would help the user understand how to handle them as well.